### DISCUSSION TITLE
How to enable custom plugin from kong manager?

### DISCUSSION THREAD

**Discussion by shahamit** [Kong/kong#9652] [Help]
We are running kong (kong/kong-gateway:3.0.0.0-alpine) with kong-manager (free mode) as docker container. We have coded a custom plugin that we do install in the running kong container through the below commands. I do not see this plugin on kong manager UI. How can I see my custom plugin and enable it for a service or route?

```
docker exec -it --user root $container mkdir /usr/local/share/lua/5.1/kong/plugins/$dir

docker cp plugin/. $container:/usr/local/share/lua/5.1/kong/plugins/$dir

docker exec --user kong -e KONG_PLUGINS="bundled,$dir" $container kong reload -vv
```
Thanks.

**Comment by hbagdi** [comment]
Do yo usee any errors in the start up logs?

**Reply by shahamit** [reply]
Thanks for the hint. It seems that the above docker commands didn't print any errors in the console but when I checked the docker logs of the gateway I could see the below error. It seems to be coming from the first line of my header.lua file
`local BasePlugin = require "kong.plugins.base_plugin"` 
When I navigated to the `/usr/local/share/lua/5.1/kong/plugins` directory in the container I do see a directory created for my custom plugin but there isn't a `base_plugin` directory. How can I access the base plugin in my code?
Thanks.
```
2022/11/03 09:03:53 [warn] 1#0: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /usr/local/kong/nginx.conf:6
2022/11/03 09:03:53 [warn] 1#0: [kong] init.lua:388 the Granular tracing is deprecated, please use OpenTelemetry tracing to track the lifecycle of Kong instead. (deprecated after 3.0.0.0, scheduled for removal in 4.0.0.0)
2022/11/03 09:03:53 [notice] 1#0: [lua] license_helpers.lua:123: read_license_info(): [license-helpers] could not decode license JSON: No license found
2022/11/03 09:03:53 [notice] 1#0: [lua] license_helpers.lua:123: read_license_info(): [license-helpers] could not decode license JSON: No license found
2022/11/03 09:03:53 [crit] 2132#0: *1869 [lua] init.lua:305: [cluster_events] no 'at' in shm, polling events from: 1667466223.954, context: ngx.timer
2022/11/03 09:03:53 [warn] 1#0: [kong] [C]:-1 [Penlight 1.13.1] the contents of module 'pl.xml' has been deprecated, please use a more specialized library instead (deprecated after 1.11.0, scheduled for removal in 2.0.0)
2022/11/03 09:03:53 [error] 1#0: init_by_lua error: /usr/local/share/lua/5.1/kong/tools/utils.lua:708: error loading module 'kong.plugins.kong-uam-plugin.handler':
...l/share/lua/5.1/kong/plugins/kong-uam-plugin/handler.lua:1: module 'kong.plugins.base_plugin' not found:No LuaRocks module found for kong.plugins.base_plugin
	no field package.preload['kong.plugins.base_plugin']
	no file './kong/plugins/base_plugin.lua'
	no file './kong/plugins/base_plugin/init.lua'
	no file '/usr/local/openresty/site/lualib/kong/plugins/base_plugin.ljbc'
	no file '/usr/local/openresty/site/lualib/kong/plugins/base_plugin/init.ljbc'
	no file '/usr/local/openresty/lualib/kong/plugins/base_plugin.ljbc'
	no file '/usr/local/openresty/lualib/kong/plugins/base_plugin/init.ljbc'
	no file '/usr/local/openresty/site/lualib/kong/plugins/base_plugin.lua'
	no file '/usr/local/openresty/site/lualib/kong/plugins/base_plugin/init.lua'
	no file '/usr/local/openresty/lualib/kong/plugins/base_plugin.lua'
	no file '/usr/local/openresty/lualib/kong/plugins/base_plugin/init.lua'
	no file '/usr/local/openresty/luajit/share/luajit-2.1.0-beta3/kong/plugins/base_plugin.lua'
	no file '/usr/local/share/lua/5.1/kong/plugins/base_plugin.lua'
	no file '/usr/local/share/lua/5.1/kong/plugins/base_plugin/init.lua'
	no file '/usr/local/openresty/luajit/share/lua/5.1/kong/plugins/base_plugin.lua'
	no file '/usr/local/openresty/luajit/share/lua/5.1/kong/plugins/base_plugin/init.lua'
	no file '/home/kong/.luarocks/share/lua/5.1/kong/plugins/base_plugin.lua'
	no file '/home/kong/.luarocks/share/lua/5.1/kong/plugins/base_plugin/init.lua'
	no file '/usr/local/share/lua/5.1/kong/plugins/base_plugin.ljbc'
	no file '/usr/local/share/lua/5.1/kong/plugins/base_plugin/init.ljbc'
	no file '/usr/local/openresty/site/lualib/kong/plugins/base_plugin.so'
	no file '/usr/local/openresty/lualib/kong/plugins/base_plugin.so'
	no file './kong/plugins/base_plugin.so'
	no file '/usr/local/lib/lua/5.1/kong/plugins/base_plugin.so'
	no file '/usr/local/openresty/luajit/lib/lua/5.1/kong/plugins/base_plugin.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
	no file '/home/kong/.luarocks/lib/lua/5.1/kong/plugins/base_plugin.so'
	no file '/usr/local/openresty/site/lualib/kong.so'
	no file '/usr/local/openresty/lualib/kong.so'
	no file './kong.so'
	no file '/usr/local/lib/lua/5.1/kong.so'
	no file '/usr/local/openresty/luajit/lib/lua/5.1/kong.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
	no file '/home/kong/.luarocks/lib/lua/5.1/kong.so'
stack traceback:
	[C]: in function 'require'
	...l/share/lua/5.1/kong/plugins/kong-uam-plugin/handler.lua:1: in main chunk
	[C]: at 0x7f9de1f602d0
	[C]: in function 'xpcall'
	/usr/local/share/lua/5.1/kong/tools/utils.lua:699: in function 'load_module_if_exists'
	/usr/local/share/lua/5.1/kong/db/dao/plugins.lua:206: in function 'load_plugin_handler'
	/usr/local/share/lua/5.1/kong/db/dao/plugins.lua:313: in function 'load_plugin'
	/usr/local/share/lua/5.1/kong/db/dao/plugins.lua:365: in function 'load_plugin_schemas'
	/usr/local/share/lua/5.1/kong/init.lua:629: in function 'init'
	init_by_lua:3: in main chunk
stack traceback:
	[C]: in function 'error'
	/usr/local/share/lua/5.1/kong/tools/utils.lua:708: in function 'load_module_if_exists'
	/usr/local/share/lua/5.1/kong/db/dao/plugins.lua:206: in function 'load_plugin_handler'
	/usr/local/share/lua/5.1/kong/db/dao/plugins.lua:313: in function 'load_plugin'
	/usr/local/share/lua/5.1/kong/db/dao/plugins.lua:365: in function 'load_plugin_schemas'
	/usr/local/share/lua/5.1/kong/init.lua:629: in function 'init'
	init_by_lua:3: in main chunk
2022/11/03 09:03:57 [notice] 1#0: signal 17 (SIGCHLD) received from 2485
2022/11/03 09:03:58 [crit] 2132#0: *1815 [lua] init.lua:305: [cluster_events] no 'at' in shm, polling events from: 1667466228.967, context: ngx.timer
```

**Reply by luozhouyang** [reply]
BasePlugin API is deprecated, update your plugin code with the new API.

**Reply by shahamit** [reply]
these are the first 2 lines in the handler.lua file of my custom plugin

```
local BasePlugin = require "kong.plugins.base_plugin"
local access = require "kong.plugins.kong-uam-plugin.access"
```

What do I need to replace them with?

**Reply by luozhouyang** [reply]
Read the official document: [plugin-development](https://docs.konghq.com/gateway/latest/plugin-development/)

**Reply by shahamit** [reply]
Thanks for the link. I updated my custom plugin code to remove the references to base_plugin. I don't see any errors now in the kong gateway docker container logs. I still don't see the plugin in kong manager UI (http://localhost:8002/default/plugins/select/).

How can I verify that the plugin got added and enabled on the gateway? Any admin rest api endpoint?

Thanks.

**Reply by luozhouyang** [reply]
You should specify your plugin name in a kong confiuration file like `kong.conf`, and then restart kong `kong restart -c /path/to/kong.conf`. Read the documentation carefully.

**Reply by shahamit** [reply]
@luozhouyang - that's already done as part of the last docker command. I do see in the logs printed on executing the `reload -vv` command the plugins variable printed `plugins = {"bundled","kong-uam-plugin"}`

Would this plugin directly show up in the Kong manager? I doubt since it would need a plugin-logo and description as other bundled plugins show up on this page (http://localhost:8002/default/plugins/select/).

**Reply by mayocream** [reply]
Hi @shahamit, AFAIK, custom plugins are manageable in Kong Manager, but yes, the missing plugin logo and description would cause some pages to show errors.

**Reply by shahamit** [reply]
@mayocream - there are no errors in the install plugin page of the kong manager UI.

**Reply by shahamit** [reply]
@luozhouyang @mayocream - any inputs on whats going wrong?

**Reply by sumimakito** [reply]
@shahamit Hello. I tested with a simple custom plugin and the steps provided in the description. However, I cannot reproduce the issue. You may check if your custom plugin is available at http://localhost:8001. Your `kong-uam-plugin` needs to appear under `plugins.available_on_server` before it can be configured via Kong Manager. Could you see your custom plugin in the response?

**Reply by shahamit** [reply]
@sumimakito - No my plugin doesn't show up on the in the api response. One point I noticed that the existing plugins enabled on kong under the folder `/usr/local/share/lua/5.1/kong/plugins/` of the kong-gateway docker container have a different file extension - handler.ljbc and schema.ljbc while I have copied my custom plugin directory as is from the source to this directory as `/usr/local/share/lua/5.1/kong/plugins/kong-uam-plugin`. This directory has handler.lua, access.lua and schema.lua files. Do the lua sources files need to be converted into .ljbc file format in some way?

**Reply by shahamit** [reply]
@sumimakito @luozhouyang @mayocream - After reverting back to kong 2.8 I could see the plugin in the api response for GET http://localhost:8001/. Its show under `configurations.plugins` and `configurations.loaded_plugins`. I don't see it under `plugins.available_on_server` yet. **I can add the plugin to a specific service or route through the admin api**. What do i need to do to see it under under kong manager?
